builder/main.py

```diff
@@ -163,7 +163,10 @@ def _parseSdccFlags(flags):
 
 # custom upload tool
 elif upload_protocol == "custom":
-    upload_actions = [env.VerboseAction("$UPLOADCMD", "Uploading $SOURCE")]
+    upload_actions = [
+        env.VerboseAction(env.AutodetectUploadPort,
+                          "Looking for upload port..."),
+        env.VerboseAction("$UPLOADCMD", "Uploading $SOURCE")]
```

Comment:
This is the custom upload method and you have full control on top of it. Developers can use different upload tools that do not depend on the Serial port. 

You can check an example about using advanced scripting and dynamic target https://docs.platformio.org/en/latest/scripting/examples/custom_upload_tool.html

Reply:
This update does not customize the uploading method, it just making the $UPLOAD_PORT variable available to the developer and still he can override it.  Otherwise, the developer need every time to update the port manually using custom uploading tool, as the original build code doesn't provide this variable like other platforms.

Reply:
1. Do we have upload tools that do not require an upload port? What do I mean, if someone will try to use a custom tool that flashes the device directly (not using bootloader), PlatformIO will require an upload port.
2. What is your custom upload tool/command?